scripts/gen_dockerfile_test.py

```diff
@@ -0,0 +1,220 @@
+     'FROM my_base_runtime_image'),
+    # Python version
+    (_base._replace(dockerfile_python_version='_my_version'), True,
+     'python_version=python_my_version'),
+])
+def test_generate_files(app_config, should_find, test_string):
+    result = gen_dockerfile.generate_files(app_config)
+    assert sorted(result.keys()) == sorted(EXPECTED_OUTPUT_FILES)
+    dockerfile = result['Dockerfile']
+    if should_find:
+        assert test_string in dockerfile
+    else:
+        assert test_string not in dockerfile
+
+
+@pytest.mark.parametrize('app', [
+    # Sample from https://github.com/GoogleCloudPlatform/python-docs-samples
+    'hello_world',
+])
+def test_generate_dockerfile_command(tmpdir, testdata_dir, app):
+    """Generates output and compares against a set of golden files.
+
+    Optionally runs 'gcloud app gen-config' and compares against that.
+    """
+    app_dir = os.path.join(testdata_dir, app)
+    temp_dir = os.path.join(str(tmpdir), app)
+    os.mkdir(temp_dir)
+
+    # Copy sample app to writable temp dir, and generate Dockerfile.
+    config_dir = os.path.join(temp_dir, 'config')
+    shutil.copytree(app_dir, config_dir)
+    gen_dockerfile.generate_dockerfile_command(
+        base_image='gcr.io/google-appengine/python',
+        config_file=os.path.join(config_dir, 'app.yaml'),
+        source_dir=config_dir)
+
+    # Compare against golden files
+    golden_dir = os.path.join(testdata_dir, app + '_golden')
+    for filename in EXPECTED_OUTPUT_FILES:
+        compare_file(filename, config_dir, golden_dir)
+
+    # Copy sample app to different writable temp dir, and
+    # generate Dockerfile using gcloud.
+    if not shutil.which('gcloud'):
+        print('"gcloud" tool not found in $PATH, skipping rest of test')
+        return
+    gen_config_dir = os.path.join(temp_dir, 'gen_config')
+    shutil.copytree(app_dir, gen_config_dir)
+    app_yaml = os.path.join(gen_config_dir, 'app.yaml')
+    gcloud_args = [
+        'gcloud', '--quiet', 'beta', 'app', 'gen-config',
+        gen_config_dir, '--custom', '--config={}'.format(app_yaml)
+    ]
+    print('Invoking gcloud as {}'.format(gcloud_args))
+    subprocess.check_call(gcloud_args)
+    for filename in EXPECTED_OUTPUT_FILES:
+        compare_file(filename, config_dir, gen_config_dir)
+
+
+@pytest.mark.parametrize('argv', [
+    [],
+    ['argv0', '--base-image=nocolon'],
+    ['argv0', '--base-image=name:andcolon'],
+    ['argv0', '--base-image=name@sha256:digest'],
+])
+def test_parse_args_valid(argv):
+    args = gen_dockerfile.parse_args(argv)
+    assert args is not None
+
+
+@pytest.mark.parametrize('argv', [
+    ['argv0', '--base-image='],
+    ['argv0', '--base-image=:'],
+    ['argv0', '--base-image=:noname'],
+])
+def test_parse_args_invalid(argv):
+    def mock_error(*args):
+        """Prevent argparse from calling sys.exit()"""
+        raise AssertionError(*args)
+
+    with unittest.mock.patch.object(argparse.ArgumentParser, 'error',
```

Comment:
To avoid awkward line continuations with mock, I prefer:

```python
error_patch = unittest.mock.patch.object(
   argparse.ArgumentParser, 'error', mock_error)

with error_patch:
    ...
```

Also, in this case I would just suggest patching `sys.exit`.